packages/jest-mock/__typetests__/mock-functions.test.ts

```diff
@@ -25,545 +20,665 @@ import {
-expectType<SpiedFunction<typeof indexSpiedObject.methodB>>(
-  spyOn(indexSpiedObject, 'methodB'),
-);
-expectType<SpiedFunction<typeof indexSpiedObject.methodC>>(
-  spyOn(indexSpiedObject, 'methodC'),
-);
-expectType<SpiedFunction<typeof indexSpiedObject.methodE>>(
-  spyOn(indexSpiedObject, 'methodE'),
-);
-
-expectType<SpiedGetter<typeof indexSpiedObject.propertyA>>(
-  spyOn(indexSpiedObject, 'propertyA', 'get'),
-);
-expectType<SpiedSetter<typeof indexSpiedObject.propertyA>>(
-  spyOn(indexSpiedObject, 'propertyA', 'set'),
-);
-expectError(spyOn(indexSpiedObject, 'propertyA'));
-
-expectError(spyOn(indexSpiedObject, 'notThere'));
-
-// interface with optional properties
-
-class SomeClass {
-  constructor(one: string, two?: boolean) {}
-
-  methodA() {
-    return true;
+describe('jest.fn()', () => {
+  const mockFnImpl: (this: Date, a: string, b?: number) => boolean = (a, b) =>
+    true;
+
+  const mockFn = fn(mockFnImpl);
+  const mockAsyncFn = fn(async (p: boolean) => 'value');
+
+  const MockObject = fn((credentials: string) => ({
+    connect() {
+      return fn();
+    },
+    disconnect() {
+      return;
+    },
+  }));
+
+  test('when sync function is provided, returned object can be chained', () => {
+    expect(
+      fn(() => 'value')
+        .mockClear()
+        .mockReset()
+        .mockImplementation(() => 'value')
+        .mockImplementationOnce(() => 'value')
+        .mockName('mock')
+        .mockReturnThis()
+        .mockReturnValue('value')
+        .mockReturnValueOnce('value'),
+    ).type.toEqual<Mock<() => string>>();
+
+    expect(
+      fn(() => 'value').mockReturnValue(Promise.resolve('value')),
+    ).type.toRaiseError();
+    expect(
+      fn(() => 'value').mockReturnValueOnce(Promise.resolve('value')),
+    ).type.toRaiseError();
+  });
+
+  test('when async function is provided, returned object can be chained', () => {
+    expect(
+      fn(async () => 'value')
+        .mockClear()
+        .mockReset()
+        .mockImplementation(fn(async () => 'value'))
+        .mockImplementationOnce(fn(async () => 'value'))
+        .mockName('mock')
+        .mockResolvedValue('value')
+        .mockResolvedValueOnce('value')
+        .mockRejectedValue('error')
+        .mockRejectedValueOnce('error')
+        .mockReturnThis()
+        .mockReturnValue(Promise.resolve('value'))
+        .mockReturnValueOnce(Promise.resolve('value')),
+    ).type.toEqual<Mock<() => Promise<string>>>();
+
+    expect(fn(() => 'value').mockResolvedValue('value')).type.toRaiseError();
+    expect(
+      fn(() => 'value').mockResolvedValueOnce('value'),
+    ).type.toRaiseError();
+
+    expect(fn(() => 'value').mockRejectedValue('error')).type.toRaiseError();
+    expect(
+      fn(() => 'value').mockRejectedValueOnce('error'),
+    ).type.toRaiseError();
+  });
+
+  test('models typings of mocked function', () => {
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    expect(fn()).type.toMatch<Function>();
```

Comment:
`.toMatch()` instead of assignability. This checks that `fn()` extends from the `Function` type. Perhaps `.toExtend()` would be better name for this matcher. In a way it is almost the same.